community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/FlippableIndexProxy.java

```diff
@@ -124,6 +137,64 @@ public void force() throws IOException
         }
     }
 
+    /**
+     * Acquire the {@code ReadLock} in an <i>unfair</i> way, without waiting for queued up writers.
+     * <p/>
+     * The {@link ReentrantReadWriteLock.ReadLock#tryLock() tryLock}-method of the {@code ReadLock} implementation of
+     * {@code ReentrantReadWriteLock} implements a <i>barging</i> behaviour, where if an exclusive lock is not held,
+     * the shared lock will be acquired, even if there are other threads waiting for the lock. This behaviour is
+     * regardless of whether the lock is fair or not.
+     * <p/>
+     * This allows us to avoid deadlocks where readers would wait for writers that wait for readers in critical
+     * methods.
+     * <p/>
+     * The naive way to implement this method would be:
+     * <pre><code>
+     *     if ( !lock.tryLock() ) // try to barge
+     *         lock.lock(); // fall back to normal blocking lock call
+     * </code></pre>
+     * This would however not implement the appropriate barging behaviour in a scenario like the following: Say the
+     * exclusive lock is held, and there is a queue waiting containing first a reader and then a writer, in this case
+     * the {@code tryLock()} method will return false. If the writer then finishes between the naive implementation
+     * exiting {@code tryLock()} and before entering {@code lock()} the {@code barge(...)} method would now block in
+     * the exact way we don't want it to block, with a read lock held and a writer waiting.<br/>
+     * In order to get around this situation, the implementation of this method uses a
+     * {@linkplain Lock#tryLock(long, TimeUnit) timed wait} in a retry-loop in order to ensure that we make another
+     * attempt to barge the lock at a later point.
+     * <p/>
+     * This method is written to be compatible with the signature of {@link Lock#lock()}, which is not interruptible,
+     * but implemented based on the interruptible {@link Lock#tryLock(long, TimeUnit)}, so the implementation needs to
+     * remember being interrupted, and reset the flag before exiting, so that later invocations of interruptible
+     * methods detect the interruption.
+     *
+     * @param lock a {@link java.util.concurrent.locks.ReentrantReadWriteLock.ReadLock}
+     */
+    public static void barge( Lock lock )
```

Comment:
It could be in a dedicated class in `org.neo4j.kernel.impl.util.concurrent`, though I also like that it's currently close to the only place where it's used. Since we are relying on implementation specific behaviour, I think the type of the `lock` field should be changed to the concrete `ReentrantReadWriteLock` type, instead of the interface.